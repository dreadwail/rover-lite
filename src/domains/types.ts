import { Action, AnyAction } from 'redux';
import { ThunkAction, ThunkDispatch } from 'redux-thunk';

import { RoverSearchState } from '../state/types';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type GenericThunkAction<T, S extends RoverSearchState = RoverSearchState> = ThunkAction<
  T,
  S,
  {},
  AnyAction
>;
export type PromiseThunkAction<
  T,
  S extends RoverSearchState = RoverSearchState
> = GenericThunkAction<Promise<T>, S>;

export type MapDispatchToProps<
  TDispatchProps,
  TOwnProps = {},
  TState extends RoverSearchState = RoverSearchState,
  A extends Action = Action
> = <E>(dispatch: ThunkDispatch<TState, E, A>, ownProps: TOwnProps) => TDispatchProps;
