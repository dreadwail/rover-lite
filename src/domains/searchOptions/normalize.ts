import {
  BasicOption,
  BooleanOption,
  ChoiceOption,
  DateOption,
  IntegerOption,
  MultipleChoiceOption,
  Option,
  Options,
} from '../../state/searchOptions/types';

interface RawBasicOption {
  readonly type: string;
  readonly required: boolean;
  readonly label: string;
  readonly help_text?: string;
  readonly category?: string;
  readonly category_rank?: number;
  readonly category_display_name?: string;
  readonly rank?: number;
}

interface RawDateOption extends RawBasicOption {
  readonly type: 'date';
}

interface RawChoice {
  readonly value: any;
  readonly display_name: string;
}

interface RawChoiceOption extends RawBasicOption {
  readonly type: 'choice';
  readonly choices: RawChoice[];
}

interface RawMultipleChoiceOption extends RawBasicOption {
  readonly type: 'multiple choice';
  readonly choices: RawChoice[];
}

interface RawIntegerOption extends RawBasicOption {
  readonly type: 'integer';
  readonly min_value: number;
  readonly max_value: number;
}

interface RawBooleanOption extends RawBasicOption {
  readonly type: 'boolean';
}

type RawOption =
  | RawBasicOption
  | RawDateOption
  | RawChoiceOption
  | RawMultipleChoiceOption
  | RawIntegerOption
  | RawBooleanOption;

const isRawBooleanOption = (rawOption: RawOption): rawOption is RawBooleanOption =>
  rawOption.type === 'boolean';
const isRawIntegerOption = (rawOption: RawOption): rawOption is RawIntegerOption =>
  rawOption.type === 'integer';
const isRawMultipleChoiceOption = (rawOption: RawOption): rawOption is RawMultipleChoiceOption =>
  rawOption.type === 'multiple choice';
const isRawChoiceOption = (rawOption: RawOption): rawOption is RawChoiceOption =>
  rawOption.type === 'choice';
const isRawDateOption = (rawOption: RawOption): rawOption is RawDateOption =>
  rawOption.type === 'date';

type OptionsResponseActionType = 'GET';

type RawOptions = Record<string, RawOption>;

export interface OptionsResponse {
  readonly name: string;
  readonly description: string;
  readonly renders: string[];
  readonly parses: string[];
  readonly actions: Record<OptionsResponseActionType, RawOptions>;
}

const extractRawCategory = (rawBasicOption: RawBasicOption): Option['category'] => {
  const defaultCategory: Option['category'] = {
    categoryId: 'detail',
    rank: 0,
    displayName: 'Detail',
  };
  if (!rawBasicOption.category) {
    return defaultCategory;
  }
  if (!rawBasicOption.category_display_name) {
    return defaultCategory;
  }
  if (!rawBasicOption.category_rank) {
    return defaultCategory;
  }
  return {
    categoryId: rawBasicOption.category,
    rank: rawBasicOption.category_rank,
    displayName: rawBasicOption.category_display_name,
  };
};

const labelOverrides: Record<string, string> = {
  Minprice: 'Minimum Price',
  Maxprice: 'Maximum Price',
};

const convertOverridableLabel = (rawLabel: string): string => {
  const override = labelOverrides[rawLabel];
  if (override) {
    return override;
  }
  return rawLabel;
};

const rawBasicOptionToOption = (rawBasicOption: RawBasicOption): BasicOption => ({
  type: rawBasicOption.type,
  required: !!rawBasicOption.required,
  label: convertOverridableLabel(rawBasicOption.label),
  helpText: rawBasicOption.help_text,
  rank: rawBasicOption.rank || 0,
  category: extractRawCategory(rawBasicOption),
});

const rawBooleanOptionToOption = (rawBooleanOption: RawBooleanOption): BooleanOption => {
  const option = rawBasicOptionToOption(rawBooleanOption);
  return { ...option, type: 'boolean' };
};

const rawIntegerOptionToOption = (rawIntegerOption: RawIntegerOption): IntegerOption => {
  const option = rawBasicOptionToOption(rawIntegerOption);
  return {
    ...option,
    type: 'integer',
    minValue: rawIntegerOption.min_value,
    maxValue: rawIntegerOption.max_value,
  };
};

const choiceDisplayNameOverrides: Record<string, string> = {
  no_pref: 'No preference',
  no_other_dogs: 'No other dogs',
  other_dogs: 'Other dogs',
  houses: 'Houses',
  apartments: 'Apartments',
};

const convertOverridableChoiceDisplayName = (rawDisplayName: string): string => {
  const override = choiceDisplayNameOverrides[rawDisplayName];
  if (override) {
    return override;
  }
  return rawDisplayName;
};

const rawChoiceOptionToOption = (rawChoiceOption: RawChoiceOption): ChoiceOption => {
  const option = rawBasicOptionToOption(rawChoiceOption);
  return {
    ...option,
    type: 'choice',
    choices: rawChoiceOption.choices.map(choice => ({
      value: choice.value,
      displayName: convertOverridableChoiceDisplayName(choice.display_name),
    })),
  };
};

const rawMultipleChoiceOptionToOption = (
  rawMultipleChoiceOption: RawMultipleChoiceOption
): MultipleChoiceOption => {
  const option = rawBasicOptionToOption(rawMultipleChoiceOption);
  return {
    ...option,
    type: 'multiple choice',
    choices: rawMultipleChoiceOption.choices.map(choice => ({
      value: choice.value,
      displayName: convertOverridableChoiceDisplayName(choice.display_name),
    })),
  };
};

const rawDateOptionToOption = (rawDateOption: RawDateOption): DateOption => {
  const option = rawBasicOptionToOption(rawDateOption);
  return { ...option, type: 'date' };
};

const rawOptionToOption = (rawOption: RawOption): Option => {
  if (isRawBooleanOption(rawOption)) {
    return rawBooleanOptionToOption(rawOption);
  }
  if (isRawIntegerOption(rawOption)) {
    return rawIntegerOptionToOption(rawOption);
  }
  if (isRawMultipleChoiceOption(rawOption)) {
    return rawMultipleChoiceOptionToOption(rawOption);
  }
  if (isRawChoiceOption(rawOption)) {
    return rawChoiceOptionToOption(rawOption);
  }
  if (isRawDateOption(rawOption)) {
    return rawDateOptionToOption(rawOption);
  }
  return rawBasicOptionToOption(rawOption);
};

const keysToSkip: string[] = ['service_type', 'page'];

export const normalize = (response: OptionsResponse): Options => {
  const rawOptions = response.actions.GET;
  return Object.keys(rawOptions).reduce((options, key) => {
    if (keysToSkip.includes(key)) {
      return options;
    }
    const rawOption = rawOptions[key];
    const option = rawOptionToOption(rawOption);
    return { ...options, [key]: option };
  }, {});
};
