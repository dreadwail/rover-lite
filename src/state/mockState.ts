import { RoverSearchState } from './types';

type Primitive = boolean | number | string;

export type RecursivePartial<ThisLevel> = {
  [Key in keyof ThisLevel]?: ThisLevel[Key] extends Primitive
    ? ThisLevel[Key]
    : RecursivePartial<ThisLevel[Key]>
};

export const createMockState = (
  interestingStateKeys?: RecursivePartial<RoverSearchState>
): RoverSearchState => (interestingStateKeys as unknown) as RoverSearchState;
