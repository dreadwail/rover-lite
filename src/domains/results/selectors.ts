import { Listing } from '../../state/results/types';
import { RoverSearchState } from '../../state/types';

export const isFetching = (state: RoverSearchState): boolean => state.results.isFetching;
export const hasFailed = (state: RoverSearchState): boolean => state.results.hasFailed;
export const isStale = (state: RoverSearchState): boolean => state.results.isStale;
export const getListings = (state: RoverSearchState): Listing[] => state.results.listings;
