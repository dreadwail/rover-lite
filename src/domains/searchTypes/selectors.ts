import { SearchType } from '../../state/searchTypes/types';
import { RoverSearchState } from '../../state/types';

export const getSearchTypes = (state: RoverSearchState): SearchType[] => {
  const endpoints = state.searchTypes.endpoints;
  return Object.keys(endpoints) as SearchType[];
};
export const hasFailed = (state: RoverSearchState): boolean => state.searchTypes.hasFailed;
export const isFetching = (state: RoverSearchState): boolean => state.searchTypes.isFetching;
export const getEndpointUrl = (
  state: RoverSearchState,
  searchType: SearchType
): string | undefined => state.searchTypes.endpoints[searchType];
export const isSearchTypesLoaded = (state: RoverSearchState): boolean =>
  getSearchTypes(state).length > 0;
