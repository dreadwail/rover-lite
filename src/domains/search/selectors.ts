import { SearchInputs } from '../../state/search/types';
import { SearchType } from '../../state/searchTypes/types';
import { RoverSearchState } from '../../state/types';

export const getSearchType = (state: RoverSearchState): SearchType => state.search.searchType;
export const getInputs = (state: RoverSearchState): SearchInputs => state.search.inputs;
export const getLatLong = (state: RoverSearchState): [number, number] => [
  state.search.latitude,
  state.search.longitude,
];
