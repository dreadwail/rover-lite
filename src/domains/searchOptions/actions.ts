import { Action } from 'redux';

import { request } from '../../api';
import { Options } from '../../state/searchOptions/types';
import { SearchType } from '../../state/searchTypes/types';
import { getEndpointUrl } from '../searchTypes/selectors';
import { PromiseThunkAction } from '../types';

import { normalize, OptionsResponse } from './normalize';

export interface FetchSearchOptionsPendingAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.PENDING';
  readonly payload: {
    readonly searchType: SearchType;
  };
}

export interface FetchSearchOptionsFulfilledAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.FULFILLED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly options: Options;
  };
}

export interface FetchSearchOptionsRejectedAction extends Action {
  readonly type: 'FETCH_SEARCH_OPTIONS.REJECTED';
  readonly payload: {
    readonly searchType: SearchType;
    readonly error: any;
  };
  readonly error: true;
}

const fetchSearchOptionsPending = (searchType: SearchType): FetchSearchOptionsPendingAction => ({
  type: 'FETCH_SEARCH_OPTIONS.PENDING',
  payload: {
    searchType,
  },
});

const fetchSearchOptionsFulfilled = (
  searchType: SearchType,
  options: Options
): FetchSearchOptionsFulfilledAction => ({
  type: 'FETCH_SEARCH_OPTIONS.FULFILLED',
  payload: {
    searchType,
    options,
  },
});

const fetchSearchOptionsRejected = (
  searchType: SearchType,
  error: any
): FetchSearchOptionsRejectedAction => ({
  type: 'FETCH_SEARCH_OPTIONS.REJECTED',
  payload: {
    searchType,
    error,
  },
  error: true,
});

export type FetchSearchOptions = (searchType: SearchType) => PromiseThunkAction<void>;
export const fetchSearchOptions: FetchSearchOptions = searchType => async (dispatch, getState) => {
  dispatch(fetchSearchOptionsPending(searchType));
  try {
    const state = getState();
    const endpointUrl = getEndpointUrl(state, searchType);
    if (!endpointUrl) {
      throw new Error(`Unhandled search type: ${searchType}`);
    }
    const response = await request<OptionsResponse>('OPTIONS', endpointUrl);
    const normalizedOptions = normalize(response);
    dispatch(fetchSearchOptionsFulfilled(searchType, normalizedOptions));
  } catch (err) {
    dispatch(fetchSearchOptionsRejected(searchType, err));
  }
};
