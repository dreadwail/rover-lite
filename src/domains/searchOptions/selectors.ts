import { Options, OptionsByCategoryId, OptionCategory } from '../../state/searchOptions/types';
import { RoverSearchState } from '../../state/types';

export const hasFailed = (state: RoverSearchState): boolean => state.searchOptions.hasFailed;
export const isFetching = (state: RoverSearchState): boolean => state.searchOptions.isFetching;

const getOptions = (state: RoverSearchState): Options => state.searchOptions.options;

export const getRankedCategories = (state: RoverSearchState): OptionCategory[] => {
  const options = getOptions(state);
  const categoriesByKey = Object.keys(options).reduce<Record<string, OptionCategory>>(
    (categories, key) => {
      const { category } = options[key];
      return { ...categories, [category.categoryId]: category };
    },
    {}
  );
  return Object.values(categoriesByKey).sort((a, b) => a.rank - b.rank);
};

export const getOptionsByCategoryId = (state: RoverSearchState): OptionsByCategoryId => {
  const options = getOptions(state);
  return Object.keys(options).reduce<OptionsByCategoryId>((byCategoryId, key) => {
    const option = options[key];
    const { category } = option;
    const optionsHere = byCategoryId[category.categoryId];
    const newOptionsHere = { ...optionsHere, [key]: option };
    return {
      ...byCategoryId,
      [category.categoryId]: newOptionsHere,
    };
  }, {});
};
