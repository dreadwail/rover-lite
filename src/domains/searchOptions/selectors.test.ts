import { createMockState } from '../../state/mockState';
import { Options, OptionCategory, SearchOptions } from '../../state/searchOptions/types';
import { RoverSearchState } from '../../state/types';

import { getOptionsByCategoryId, getRankedCategories, hasFailed, isFetching } from './selectors';

describe('search options selectors', () => {
  describe('.getRankedCategories', () => {
    let selected: OptionCategory[];

    describe('when there are no options', () => {
      beforeEach(() => {
        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: false,
          options: {},
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = getRankedCategories(state);
      });

      it('returns an empty array', () => {
        expect(selected).toEqual([]);
      });
    });

    describe('when there are options', () => {
      beforeEach(() => {
        const options: Options = {
          start_date: {
            type: 'date',
            required: false,
            label: 'Drop Off',
            helpText: "The date of the drop-off at the sitter's location",
            category: {
              categoryId: 'service_details',
              rank: 9,
              displayName: 'Service Details',
            },
            rank: 42,
          },
          end_date: {
            type: 'date',
            required: false,
            label: 'Pick Up',
            helpText: "The date of the pick-up from the sitter's location",
            category: {
              categoryId: 'provider_skills',
              rank: 2,
              displayName: 'Provider Skills',
            },
            rank: 2,
          },
        };

        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: false,
          options,
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = getRankedCategories(state);
      });

      it('returns the categories ordered by their rank', () => {
        expect(selected).toMatchSnapshot();
      });
    });
  });

  describe('.getOptionsByCategoryId', () => {
    let selected: Record<string, Options>;

    describe('when there are no search options', () => {
      beforeEach(() => {
        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: false,
          options: {},
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = getOptionsByCategoryId(state);
      });

      it('returns an empty object', () => {
        expect(selected).toEqual({});
      });
    });

    describe('when there are options', () => {
      beforeEach(() => {
        const options: Options = {
          start_date: {
            type: 'date',
            required: false,
            label: 'Drop Off',
            helpText: "The date of the drop-off at the sitter's location",
            category: {
              categoryId: 'service_details',
              rank: 1,
              displayName: 'Service Details',
            },
            rank: 1,
          },
          end_date: {
            type: 'date',
            required: false,
            label: 'Pick Up',
            helpText: "The date of the pick-up from the sitter's location",
            category: {
              categoryId: 'provider_skills',
              rank: 2,
              displayName: 'Provider Skills',
            },
            rank: 2,
          },
        };

        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: false,
          options,
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = getOptionsByCategoryId(state);
      });

      it('returns the expected options', () => {
        expect(selected).toMatchSnapshot();
      });
    });
  });

  describe('.hasFailed', () => {
    let selected: boolean;

    describe('when fetching search options has failed', () => {
      beforeEach(() => {
        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: true,
          options: {},
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = hasFailed(state);
      });

      it('returns true', () => {
        expect(selected).toBe(true);
      });
    });

    describe('when fetching search types has not failed', () => {
      beforeEach(() => {
        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: false,
          options: {},
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = hasFailed(state);
      });

      it('returns false', () => {
        expect(selected).toBe(false);
      });
    });
  });

  describe('.isFetching', () => {
    let selected: boolean;

    describe('when fetching search options', () => {
      beforeEach(() => {
        const searchOptions: SearchOptions = {
          isFetching: true,
          hasFailed: false,
          options: {},
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = isFetching(state);
      });

      it('returns true', () => {
        expect(selected).toBe(true);
      });
    });

    describe('when not fetching search options', () => {
      beforeEach(() => {
        const searchOptions: SearchOptions = {
          isFetching: false,
          hasFailed: false,
          options: {},
        };
        const state: RoverSearchState = createMockState({ searchOptions });
        selected = isFetching(state);
      });

      it('returns false', () => {
        expect(selected).toBe(false);
      });
    });
  });
});
