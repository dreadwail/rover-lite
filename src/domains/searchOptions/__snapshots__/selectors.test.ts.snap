// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`search options selectors .getOptionsByCategoryId when there are options returns the expected options 1`] = `
Object {
  "provider_skills": Object {
    "end_date": Object {
      "category": Object {
        "categoryId": "provider_skills",
        "displayName": "Provider Skills",
        "rank": 2,
      },
      "helpText": "The date of the pick-up from the sitter's location",
      "label": "Pick Up",
      "rank": 2,
      "required": false,
      "type": "date",
    },
  },
  "service_details": Object {
    "start_date": Object {
      "category": Object {
        "categoryId": "service_details",
        "displayName": "Service Details",
        "rank": 1,
      },
      "helpText": "The date of the drop-off at the sitter's location",
      "label": "Drop Off",
      "rank": 1,
      "required": false,
      "type": "date",
    },
  },
}
`;

exports[`search options selectors .getRankedCategories when there are options returns the categories ordered by their rank 1`] = `
Array [
  Object {
    "categoryId": "provider_skills",
    "displayName": "Provider Skills",
    "rank": 2,
  },
  Object {
    "categoryId": "service_details",
    "displayName": "Service Details",
    "rank": 9,
  },
]
`;
